### PR TITLE
ARROW-659: [C++] Add multithreaded memcpy implementation

### PR DESCRIPTION
**Pull request by atumanov**
parallelize memcopy operations for large objects with a multi-threaded implementation.

### CODE REVIEW

**Comment by pcmoritz** [comment]
@wesm please let me know if you have an idea where the best place to link against pthread in the CMakeLists.txt is (see the error in https://travis-ci.org/apache/arrow/jobs/224624635)

File: cpp/src/arrow/io/memory.h [hunk 1]
```diff
@@ -65,14 +107,16 @@ class ARROW_EXPORT BufferOutputStream : public OutputStream {
...
 };
 
 /// \brief Enables random writes into a fixed-size mutable buffer
 ///
 class ARROW_EXPORT FixedSizeBufferWriter : public WriteableFile {
  public:
   /// Input buffer must be mutable, will abort if not
-  explicit FixedSizeBufferWriter(const std::shared_ptr<Buffer>& buffer);
+  explicit FixedSizeBufferWriter(const std::shared_ptr<Buffer>& buffer,
+    std::unique_ptr<Memcopy> memcopy = nullptr);
```
Comment (wesm) [inline cpp/src/arrow/io/memory.h]: What do you think about instead having:

```
writer->set_memcpy_threads(nthreads)
writer->set_memcpy_blocksize(blocksize)
```

And then you don't need the classes.

File: cpp/src/arrow/io/memory.cc [hunk 2]
```diff
@@ -33,16 +33,67 @@
 namespace arrow {
 namespace io {
 
+void SerialMemcopy::memcopy(uint8_t* dst, const uint8_t* src, uint64_t nbytes) {
+  std::memcpy(dst, src, nbytes);
+}
+
+void ParallelMemcopy::memcopy(uint8_t* dst, const uint8_t* src, uint64_t nbytes) {
+  if (nbytes >= BYTES_IN_MB) {
```
Comment (wesm) [inline cpp/src/arrow/io/memory.cc]: Can you use google naming style on this constant `kBytesInMegabyte`

File: cpp/src/arrow/io/memory.h [hunk 3]
```diff
@@ -24,12 +24,16 @@
 #include <memory>
 #include <mutex>
 #include <string>
+#include <thread>
+#include <vector>
```
Comment (wesm) [inline cpp/src/arrow/io/memory.h]: Per comments below, if you allow setting parameters in the writer class, then we only need a couple of helper functions.

File: cpp/src/arrow/io/memory.cc [hunk 4]
```diff
@@ -33,16 +33,67 @@
...
+        reinterpret_cast<uint8_t*>(left_address) + i * chunk_size, chunk_size);
+  }
+
+  memcpy(dst, src, prefix);
+  memcpy(dst + prefix + num_threads * chunk_size,
+      reinterpret_cast<uint8_t*>(right_address), suffix);
+  for (auto& t : threadpool_) {
+    if (t.joinable()) { t.join(); }
+  }
+}
```
Comment (wesm) [inline cpp/src/arrow/io/memory.cc]: Would it be better to put this code in a utility header under `arrow/util`?

File: cpp/src/arrow/io/memory.cc [hunk 5]
```diff
@@ -33,16 +33,67 @@
...
 
+void SerialMemcopy::memcopy(uint8_t* dst, const uint8_t* src, uint64_t nbytes) {
+  std::memcpy(dst, src, nbytes);
+}
+
+void ParallelMemcopy::memcopy(uint8_t* dst, const uint8_t* src, uint64_t nbytes) {
+  if (nbytes >= BYTES_IN_MB) {
+    memcopy_aligned(dst, src, nbytes, block_size_);
+  } else {
+    memcpy(dst, src, nbytes);
```
Comment (wesm) [inline cpp/src/arrow/io/memory.cc]: I'm indifferent between `std::memcpy` and `memcpy` as long as we use one consistently

**Comment by wesm** [comment]
@pcmoritz I would put pthread in https://github.com/apache/arrow/blob/master/cpp/CMakeLists.txt#L880. After adding it there I have

```
$ ldd debug/libarrow.so
	linux-vdso.so.1 =>  (0x00007ffcc4459000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fe084014000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fe083e0f000)
	libstdc++.so.6 => /home/wesm/code/native-toolchain/build/gcc-4.9.2/lib64/libstdc++.so.6 (0x00007fe083b05000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fe0837ff000)
	libgcc_s.so.1 => /home/wesm/code/native-toolchain/build/gcc-4.9.2/lib64/libgcc_s.so.1 (0x00007fe0835e8000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fe083220000)
	/lib64/ld-linux-x86-64.so.2 (0x0000563edc9f9000)
```

File: cpp/src/arrow/io/memory.h (hunk 1 above)
Comment (wesm) [inline cpp/src/arrow/io/memory.h]: We could add other parameters, too, like a size threshold for parallelization (i.e. you wouldn't want to spawn a thread pool to copy 1024 bytes)

File: cpp/src/arrow/io/memory.h [hunk 6]
```diff
@@ -38,11 +42,49 @@ class Status;
 
 namespace io {
 
+class ARROW_EXPORT Memcopy {
+public:
+  virtual void memcopy(uint8_t* dst, const uint8_t* src, uint64_t nbytes) = 0;
```
Comment (wesm) [inline cpp/src/arrow/io/memory.h]: This is missing a virtual destructor and I believe will leak memory. I don't think we need these classes, though, only some configurable members on the Writer class

Comment (pcmoritz) [inline cpp/src/arrow/io/memory.h]: Yeah, I will make these changes in a bit, right now I'm still figuring out the benchmark ;)

Comment (wesm) [inline cpp/src/arrow/io/memory.h]: Yeah, gbenchmark is a bit obtuse sometimes =)

**Comment by pcmoritz** [comment]
This is now ready to merge. The speedups on amazon X1 instances is from 6.0GB/s to 8.4GB/s for large objects.
```
ubuntu@ip-XXX-XXX-XXX-XXX:~/arrow/cpp/build/release$ ./io-memory-benchmark 
Run on (128 X 1200.22 MHz CPU s)
2017-04-24 01:15:18
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
Benchmark                                                 Time           CPU Iterations
---------------------------------------------------------------------------------------
BM_SerialMemcopy/1/min_time:1.000/real_time        16301075 ns   16301488 ns         75   5.99079GB/s
BM_SerialMemcopy/4/min_time:1.000/real_time        16217594 ns   16217909 ns         75   6.02162GB/s
BM_SerialMemcopy/16/min_time:1.000/real_time       16230552 ns   16230939 ns         75   6.01682GB/s
BM_SerialMemcopy/64/min_time:1.000/real_time       16145794 ns   16146349 ns         75    6.0484GB/s
BM_SerialMemcopy/256/min_time:1.000/real_time      16603765 ns   16604095 ns         74   5.88157GB/s
BM_SerialMemcopy/1024/min_time:1.000/real_time     16083646 ns   16084107 ns         75   6.07177GB/s
BM_SerialMemcopy/4k/min_time:1.000/real_time       16149854 ns   16150257 ns         75   6.04688GB/s
BM_SerialMemcopy/8k/min_time:1.000/real_time       16142911 ns   16142800 ns         75   6.04948GB/s
BM_ParallelMemcopy/1/min_time:1.000/real_time      11634463 ns     129216 ns        100    8.3937GB/s
BM_ParallelMemcopy/4/min_time:1.000/real_time      11695138 ns     142703 ns         93   8.35016GB/s
BM_ParallelMemcopy/16/min_time:1.000/real_time     11565459 ns     138614 ns        101   8.44379GB/s
BM_ParallelMemcopy/64/min_time:1.000/real_time     12094462 ns     132504 ns        121   8.07446GB/s
BM_ParallelMemcopy/256/min_time:1.000/real_time    11564869 ns     126829 ns        110   8.44422GB/s
BM_ParallelMemcopy/1024/min_time:1.000/real_time   12148854 ns     130363 ns         95   8.03831GB/s
BM_ParallelMemcopy/4k/min_time:1.000/real_time     11609889 ns     139616 ns        105   8.41147GB/s
BM_ParallelMemcopy/8k/min_time:1.000/real_time     11591083 ns     129999 ns        110   8.42512GB/s
```
There is more potential speedup using low-level hacking which we should investigate in the future:
http://stackoverflow.com/a/26256216

**Review by wesm — APPROVED** [review]
+1, thank you! Can you remove the WIP from the PR title and I'll merge once the Appveyor build runs?

**Comment by pcmoritz** [comment]
@atumanov @wesm Can one of you remove the [WIP]? I can't do it because I didn't create the PR.

**Comment by wesm** [comment]
If you enable Appveyor on your fork, you can get more timely builds

**Comment by wesm** [comment]
Thank you for the contribution!

**Pull request closed without merging**